File: Source/Remote Control Server/MouseV3.vb
Imports System.Windows.Forms

Module MouseV3

    Private scrollAmount As Integer

    Private currentGesture As Byte = 0
    Private Const GESTURE_NONE As Byte = 0
    Private Const GESTURE_ZOOM As Byte = 1
    Private Const GESTURE_SCROLL As Byte = 2

    Private Const CLICK_OFFSET_TOLERANCE As Byte = 10 'dip
    Private Const CLICK_DELAY_TOLERANCE As Integer = 500

    Private Const SCROLL_OFFSET_TOLERANCE As Integer = 50

    Private Const MOUSE_LEFT_DOWN = &H2
    Private Const MOUSE_LEFT_UP = &H4
    Private Const MOUSE_RIGHT_DOWN = &H8
    Private Const MOUSE_RIGHT_UP = &H10
    Private Const MOUSE_WHEEL As Int32 = 2048
    Private Const MOUSE_WHEEL_DELTA As Int32 = 120

    Private cursorPositonNew, cursorPositonCurrent, cursorPositonDown As Point

    Private P_ORIGIN As Point = New Point(0, 0)
    Private P1_New, P2_New, P3_New As Point
    Private P1_Rel, P2_Rel, P3_Rel As Point
    Private P1_Start, P2_Start, P3_Start As Point
    Private P1_Last, P2_Last, P3_Last As Point
    Private P3_Vektor_New, P3_Vektor_Last, P3_Vektor_Start, p3_Vektor_Event As Decimal

    Private dateLeftDown As DateTime
    Private dateLeftUp As DateTime

    Private pointDown As TouchPoint
    Private pointers As List(Of TouchPoint)
    Private downEventIndex As Integer = -1
    Private moveEventIndex As Integer = -1

    Public mousePadDown, mouseLeftDown, mouseRightDown, isMultitouch As Boolean

    Structure PointAPI
        Public x As Int32
        Public y As Int32
    End Structure

    Public Declare Function GetCursorPos Lib "user32" (ByRef lpPoint As PointAPI) As Boolean

    Dim pointerLock As New Object


#Region "Basic actions"

    Public Sub leftMouseDown()
        mouseLeftDown = True
        mouse_event(MOUSE_LEFT_DOWN, 0, 0, 0, 0)
    End Sub

    Public Sub leftMouseUp()
        mouseLeftDown = False
        mouse_event(MOUSE_LEFT_UP, 0, 0, 0, 0)
    End Sub

    Public Sub leftClick()
        leftMouseDown()
        leftMouseUp()
    End Sub

    Public Sub rightMouseDown()
        mouseRightDown = True
        mouse_event(MOUSE_RIGHT_DOWN, 0, 0, 0, 0)
    End Sub

    Public Sub rightMouseUp()
        mouseRightDown = False
        mouse_event(MOUSE_RIGHT_UP, 0, 0, 0, 0)
    End Sub

    Public Sub rightClick()
        rightMouseDown()
        rightMouseUp()
    End Sub

#End Region

#Region "Pointer handling"

    Public Sub pointersDown()
        If mousePadDown Then
            Return
        End If

        'Logger.add("pointersDown 1")

        mousePadDown = True
        P1_Start = P_ORIGIN
        P1_Last = P_ORIGIN

        pointers = New List(Of TouchPoint)
        Dim currentPointer As New TouchPoint()
        currentPointer.x = 0
        currentPointer.y = 0
        pointers.Add(currentPointer)

        pointDown = New TouchPoint()
        pointDown.timestamp = Date.Now.Ticks / TimeSpan.TicksPerMillisecond

        'Logger.add("pointersDown 2")
    End Sub

    Public Sub pointersUp()
        'Logger.add("pointersUp 1")

        checkForClick()

        P1_Start = P_ORIGIN
        P1_Last = P_ORIGIN
        P2_Start = P_ORIGIN
        P2_Last = P_ORIGIN
        P3_Start = P_ORIGIN
        P3_Last = P_ORIGIN
        P3_Vektor_Start = 0
        P3_Vektor_Last = 0

        mousePadDown = False
        currentGesture = GESTURE_NONE
        pointers = New List(Of TouchPoint)

        'Logger.add("pointersUp 2")
    End Sub

    Public Sub parsePointerData(ByVal data As Byte())
        Try
            'Pointer data starts at byte index 3
            'Each pointer has two bytes for X and two for Y value

            Dim newDownEventIndex As Integer = data(3)
            newDownEventIndex = (newDownEventIndex << 8) + data(4)

            Dim newMoveEventIndex As Integer = data(5)
            newMoveEventIndex = (newMoveEventIndex << 8) + data(6)

            'Check if command is old and should be ignored
            If newDownEventIndex < downEventIndex Then
                'Logger.add("Ignoring old command (down index)")
                'Return
            ElseIf newDownEventIndex < downEventIndex And newMoveEventIndex < moveEventIndex Then
                'Logger.add("Ignoring old command (move index)")
                'Return
            End If

            Dim pointerDataOffset As Byte = 7
            Dim pointerLength As Byte = 4
            Dim pointerCount As Byte = (data.Length - pointerDataOffset) / pointerLength

            pointers = New List(Of TouchPoint)

            For i As Byte = 0 To pointerCount - 1 Step 1
                Dim currentPointer As New TouchPoint

                Dim x As Integer = data(pointerDataOffset + (i * pointerLength))
                x = (x << 8) + data(pointerDataOffset + (i * pointerLength) + 1)

                Dim y As Integer = data(pointerDataOffset + (i * pointerLength) + 2)
                y = (y << 8) + data(pointerDataOffset + (i * pointerLength) + 3)

                currentPointer.x = x
                currentPointer.y = y

                pointers.Add(currentPointer)
            Next

            'Logger.add("Down #" & newDownEventIndex & " Move #" & newMoveEventIndex & " " & pointers(0).x & "|" & pointers(0).y)

            If newDownEventIndex > downEventIndex Then
                'New cursor update should be processed after pointerUp and pointerDown event
                moveEventIndex = -1
                pointersUp()
            End If

            downEventIndex = newDownEventIndex
            moveEventIndex = newMoveEventIndex

            If pointers.Count = 1 Then
                updateCursorPosition()
            ElseIf pointers.Count > 1 Then
                parseMultitouch()
            End If

        Catch ex As Exception
        End Try
    End Sub

    Public Sub parseAbsolutePointerData(ByVal data As Byte(), Optional ByVal isPresenter As Boolean = False)
        'Pointer data starts at byte index 3
        'Each pointer has two bytes for X and two for Y value

        Dim absoluteMaximum As Integer = 32767
        Dim pointerDataOffset As Byte = 3
        Dim pointerCount As Byte = (data.Length - pointerDataOffset) / 4

        pointers = New List(Of TouchPoint)

        Dim currentPointer As New TouchPoint

        Dim x As Integer = data(pointerDataOffset)
        x = (x << 8) + data(pointerDataOffset + 1)

        Dim y As Integer = data(pointerDataOffset + 2)
        y = (y << 8) + data(pointerDataOffset + 3)

        Dim locations As System.Drawing.Point() = getScreenBounds(Screenshot.screenIndex)
        Dim startLocation As System.Drawing.Point = locations(0)
        Dim endLocation As System.Drawing.Point = locations(1)
        Dim screenSize As Size = New Size(endLocation.X - startLocation.X, endLocation.Y - startLocation.Y)

        currentPointer.x = (x * screenSize.Width) / absoluteMaximum
        currentPointer.y = (y * screenSize.Height) / absoluteMaximum

        cursorPositonNew.X = currentPointer.x
        cursorPositonNew.Y = currentPointer.y

        'Add absolute pointer position to pointer list to enable click detection
        pointers.Add(currentPointer)

        P1_Start.X = currentPointer.x
        P1_Start.Y = currentPointer.y

        SetCursorPos(cursorPositonNew.X, cursorPositonNew.Y)

        If isPresenter Then
            updatePointerPosition()
        Else

        End If
    End Sub

    Private Sub updatePointerPosition()
        If pointers.Count > 0 Then
            If Not Server.pointer.isVisible Then
                Server.pointer.showPointer()
            End If
            Dim locations As System.Drawing.Point() = Screenshot.getScreenBounds(Screenshot.screenIndex)

            Dim startLocation As System.Drawing.Point = locations(0)
            Dim endLocation As System.Drawing.Point = locations(1)
            Dim screenSize As Size = New Size(endLocation.X - startLocation.X, endLocation.Y - startLocation.Y)

            Dim width As Integer = screenSize.Width
            Dim height As Integer = screenSize.Height

            Dim pointerPosition As New Point
            Dim pointerOffset As Integer = 50
            pointerPosition.X = Math.Max(0, pointers(0).x - pointerOffset)
            pointerPosition.X = Math.Min(width - 100, pointerPosition.X)
            pointerPosition.Y = Math.Max(0, pointers(0).y - pointerOffset)
            pointerPosition.Y = Math.Min(height - 100, pointerPosition.Y)

            Server.pointer.setPointerPosition(pointerPosition)
            Server.pointer.fadeOutPointer()
        End If
    End Sub

    Private Sub updateCursorPosition()
        If mousePadDown = False Then
            'onMove event before onDown event received
            'Logger.add("onMove event before onDown event received")
            pointersDown()
            Return
        End If

        'User last pointer data to modify cursor position
        P1_New = New Point(pointers(0).x, pointers(0).y)

        If P1_Start = P_ORIGIN Or P1_Start = P_ORIGIN Then
            'Logger.add("P_Start & P_Last set to P_New: " & P1_New.X & "|" & P1_New.Y)
            P1_Start = P1_New
            P1_Last = P1_New
            Return
        End If

        cursorPositonCurrent = getCursorPosition()

        If Settings.mouseAcceleration = 1 Then
            'No acceleration
            P1_Rel.X = (P1_New.X - P1_Last.X) * Settings.mouseSensitivity
            P1_Rel.Y = (P1_New.Y - P1_Last.Y) * Settings.mouseSensitivity
        Else
            'Acceleration
            P1_Rel.X = (P1_New.X - P1_Last.X)
            P1_Rel.Y = (P1_New.Y - P1_Last.Y)

            If Not (P1_Rel.X = 0 And P1_Rel.Y = 0) Then
                If P1_Rel.X > 0 Then
                    P1_Rel.X = (P1_Rel.X ^ Settings.mouseAcceleration) * Settings.mouseSensitivity
                Else
                    P1_Rel.X = P1_Rel.X * (-1)
                    P1_Rel.X = (P1_Rel.X ^ Settings.mouseAcceleration)
                    P1_Rel.X = P1_Rel.X * Settings.mouseSensitivity
                    P1_Rel.X = P1_Rel.X * (-1)
                End If

                If P1_Rel.Y > 0 Then
                    P1_Rel.Y = (P1_Rel.Y ^ Settings.mouseAcceleration) * Settings.mouseSensitivity
                Else
                    P1_Rel.Y = P1_Rel.Y * (-1)
                    P1_Rel.Y = (P1_Rel.Y ^ Settings.mouseAcceleration)
                    P1_Rel.Y = P1_Rel.Y * Settings.mouseSensitivity
                    P1_Rel.Y = P1_Rel.Y * (-1)
                End If
            End If
        End If

        If P1_Rel.X < -700 Or P1_Rel.X > 700 Or P1_Rel.Y < -700 Or P1_Rel.Y > 700 Then
            'Logger.add("Rel: " & P1_Rel.X & "|" & P1_Rel.Y & " New: " & P1_New.X & "|" & P1_New.Y & " Old: " & P1_Last.X & "|" & P1_Last.Y)
            pointersUp()
            Return
        End If

        P1_Last = P1_New

        'Logger.add("Rel: " & P1_Rel.X & "|" & P1_Rel.Y)

        cursorPositonNew = cursorPositonCurrent + P1_Rel
        SetCursorPos(cursorPositonNew.X, cursorPositonNew.Y)
    End Sub

    Private Sub parseMultitouch()
        If mousePadDown = False Then
            'onMove event before onDown event received
            'Logger.add("onMove event before onDown event received")
            pointersDown()
            Return
        End If

        'User last pointer data to modify cursor position
        P1_New = New Point(pointers(0).x, pointers(0).y)
        P2_New = New Point(pointers(1).x, pointers(1).y)

        If P1_Start = P_ORIGIN Or P1_Last = P_ORIGIN Or P2_Start = P_ORIGIN Or P2_Last = P_ORIGIN Then
            'Logger.add("P_Start & P_Last set to P_New: " & P1_New.X & "|" & P1_New.Y)
            P1_Start = P1_New
            P1_Last = P1_New
            P2_Start = P2_New
            P2_Last = P2_New
            Return
        End If

        P1_Rel.X = (P1_New.X - P1_Last.X) * Settings.mouseSensitivity
        P1_Rel.Y = (P1_New.Y - P1_Last.Y) * Settings.mouseSensitivity

        If P1_Rel.X < -700 Or P1_Rel.X > 700 Or P1_Rel.Y < -700 Or P1_Rel.Y > 700 Then
            'Logger.add("Rel: " & P1_Rel.X & "|" & P1_Rel.Y & " New: " & P1_New.X & "|" & P1_New.Y & " Old: " & P1_Last.X & "|" & P1_Last.Y)
            pointersUp()
            Return
        End If

        P1_Last = P1_New
        P2_Last = P2_New

        processMultitouch()
    End Sub

    Public Sub processMultitouch()
        P3_New.X = Math.Round((P1_New.X + P2_New.X) / 2)
        P3_New.Y = Math.Round((P1_New.Y + P2_New.Y) / 2)
        P3_Vektor_New = Converter.getPointDistance(P1_New, P2_New)


        If P3_Start = P_ORIGIN Then
            P3_Start = P3_New
        End If

        If P3_Vektor_Start = 0 Then
            P3_Vektor_Start = P3_Vektor_New
        End If

        If currentGesture = GESTURE_NONE Then
            'Logger.add("P1: " & P1_New.X & "|" & P1_New.Y & " " & "P2: " & P2_New.X & "|" & P2_New.Y & " " & "P3: " & P3_New.X & "|" & P3_New.Y)
            'Logger.add("Determining gesture. Distance: " & P3_Vektor_New & " X: " & P3_New.X - P3_Start.X & " Y: " & P3_New.Y - P3_Start.Y)

            If valueMatchesTolerance(P3_Vektor_New, P3_Vektor_Start, SCROLL_OFFSET_TOLERANCE) Then
                'distance constant
                If valueMatchesTolerance(P3_New.X, P3_Start.X, SCROLL_OFFSET_TOLERANCE) Then
                    'horizontal constant
                    If valueMatchesTolerance(P3_New.Y, P3_Start.Y, SCROLL_OFFSET_TOLERANCE / 2) Then
                        'vertical constant
                    Else
                        'scroll
                        'Logger.add("Gesture set to scroll")
                        currentGesture = GESTURE_SCROLL
                    End If
                End If
            Else
                'zoom
                'Logger.add("Gesture set to zoom")
                currentGesture = GESTURE_ZOOM
                p3_Vektor_Event = P3_Vektor_New + 20
            End If
        End If

        If currentGesture = GESTURE_ZOOM Then
            'zoom
            If Not valueMatchesTolerance(P3_Vektor_New, p3_Vektor_Event, 25) Then
                scrollAmount = (P3_Vektor_New - P3_Vektor_Last) * (P3_Vektor_New - P3_Vektor_Last)
                If P3_Vektor_New > P3_Vektor_Last Then
                    zoom(1, 1)
                Else
                    'zoom out
                    zoom(-1, 1)
                End If
                p3_Vektor_Event = P3_Vektor_New
            End If

        ElseIf currentGesture = GESTURE_SCROLL Then
            'scroll
            scrollAmount = (P3_New.Y - P3_Last.Y) * (P3_New.Y - P3_Last.Y)
            If P3_New.Y > P3_Last.Y Then
                'scroll down
                scrollRepeat(-30 - scrollAmount, 1)
            ElseIf P3_New.Y < P3_Last.Y Then
                'scroll up
                scrollRepeat(30 + scrollAmount, 1)
            End If
        End If

        P3_Last = P3_New
        P3_Vektor_Last = P3_Vektor_New
    End Sub

    Private Sub checkForClick()
        Try
            'Down and Up event timestamp in a short timeframe?
            Dim timeDelta As Long = ((Date.Now.Ticks / TimeSpan.TicksPerMillisecond) - pointDown.timestamp)

            If timeDelta < CLICK_DELAY_TOLERANCE Then
                Logger.add("Click delay: " & timeDelta)
                'Down and Up event in the same location?
                Dim offsetX As Integer = Math.Sqrt((P1_Start.X - pointers(0).x) ^ 2)
                Dim offsetY As Integer = Math.Sqrt((P1_Start.Y - pointers(0).y) ^ 2)

                If offsetX < CLICK_OFFSET_TOLERANCE And offsetY < CLICK_OFFSET_TOLERANCE Then
                    leftClick()
                End If
            Else
                'Logger.add("Delta: " & timeDelta)
            End If
        Catch ex As Exception
        End Try
    End Sub

#End Region

#Region "Helper functions"

    Public Function getCursorPosition() As Point
        Dim lpPoint As PointAPI
        GetCursorPos(lpPoint)
        Dim cursorPosition As Point = New Point(lpPoint.x, lpPoint.y)
        Return cursorPosition
    End Function

    Public Function valueMatchesTolerance(ByVal val1 As Decimal, ByVal val2 As Decimal, Optional ByVal tolerance As Integer = CLICK_OFFSET_TOLERANCE) As Boolean
        If val1 < val2 + tolerance And val1 > val2 - tolerance Then
            Return True
        Else
            Return False
        End If
    End Function

#End Region


End Module

File: Source/MacOS/RemoteControlServer/mousev3.cpp
#include "converter.h"
#include "logger.h"
#include "mousev3.h"
#include "screenshot.h"
#include "server.h"
#include "settings.h"
#include "touchpoint.h"

#ifdef Q_OS_MAC
    #include "keyboardmac.h"
    #include "mousemac.h"
#endif

#include <QDateTime>
#include <QList>

#include "math.h"

#include <QDebug>

MouseV3* MouseV3::instance = NULL;

MouseV3* MouseV3::Instance()
{
    if (!instance)
    {
        instance = new MouseV3();
    }
    return instance;
}

MouseV3::MouseV3()
{
    cursorPositionNew = new QPoint();
    P_ORIGIN = new QPoint(0, 0);
    P1_New = new QPoint();
    P2_New = new QPoint();
    P3_New = new QPoint();
    P1_Rel = new QPoint();
    P1_Start = new QPoint();
    P2_Start = new QPoint();
    P3_Start = new QPoint();
    P1_Last = new QPoint();
    P2_Last = new QPoint();
    P3_Last = new QPoint();
    downEventIndex = -1;
    moveEventIndex = -1;

    pointDown = new TouchPoint();
}

void MouseV3::leftSecondClick()
{
    MouseMac::Instance()->leftMouseDown(true);
    MouseMac::Instance()->leftMouseUp(true);
}

void MouseV3::pointersDown()
{
    if (mousePadDown)
    {
        return;
    }

    mousePadDown = true;
    P1_Start = new QPoint(P_ORIGIN->x(), P_ORIGIN->y());
    P1_Last = new QPoint(P_ORIGIN->x(), P_ORIGIN->y());

    pointers = new QList<TouchPoint*>();
    TouchPoint* currentPointer = new TouchPoint();
    currentPointer->x = 0;
    currentPointer->y = 0;
    pointers->append(currentPointer);

    pointDown = new TouchPoint();
    pointDown->timestamp = QDateTime::currentDateTime().toMSecsSinceEpoch();
}

void MouseV3::pointersUp()
{
    checkForClick();
    P1_Start = new QPoint(P_ORIGIN->x(), P_ORIGIN->y());
    P1_Last = new QPoint(P_ORIGIN->x(), P_ORIGIN->y());
    P2_Start = new QPoint(P_ORIGIN->x(), P_ORIGIN->y());
    P2_Last = new QPoint(P_ORIGIN->x(), P_ORIGIN->y());
    P3_Start = new QPoint(P_ORIGIN->x(), P_ORIGIN->y());
    P3_Last = new QPoint(P_ORIGIN->x(), P_ORIGIN->y());
    P3_Vector_Start = 0;
    P3_Vector_Last = 0;

    mousePadDown = false;
    currentGesture = GESTURE_NONE;
    pointers = new QList<TouchPoint*>();
}

void MouseV3::parsePointerData(QByteArray &data)
{
    if (data.length() >= 7)
    {
        unsigned char firstByte = data.at(3);
        unsigned char secondByte = data.at(4);
        int newDownEventIndex = firstByte << 8 | secondByte;
        firstByte = data.at(5);
        secondByte = data.at(6);
        int newMoveEventIndex = firstByte << 8 | secondByte;

        int pointerDataOffset = 7;
        int pointerLength = 4;
        int pointerCount = (data.length() - pointerDataOffset) / pointerLength;

        pointers = new QList<TouchPoint*>();

        for (int i = 0; i < pointerCount; ++i)
        {
            TouchPoint *currentPointer = new TouchPoint();
            firstByte = data.at(pointerDataOffset + (i * pointerLength));
            secondByte = data.at(pointerDataOffset + (i * pointerLength) + 1);
            int x = firstByte << 8 | secondByte;

            firstByte = data.at(pointerDataOffset + (i * pointerLength) + 2);
            secondByte = data.at(pointerDataOffset + (i * pointerLength) + 3);
            int y = firstByte << 8 | secondByte;

            currentPointer->x = x;
            currentPointer->y = y;

            pointers->append(currentPointer);
        }

        if (newDownEventIndex > downEventIndex)
        {
            moveEventIndex = -1;
            pointersUp();
        }

        downEventIndex = newDownEventIndex;
        moveEventIndex = newMoveEventIndex;

        if (pointers->length() == 1)
        {
            updateCursorPosition();
        }
        else if (pointers->length() > 1)
        {
            parseMultitouch();
        }
    }
    else
    {
        qDebug() << "Not enough data for parsePointerData";
    }
}

void MouseV3::parseAbsolutePointerData(QByteArray &data, bool isPresenter = false)
{
    int absoluteMaximum = 32767;
    int pointerDataOffset = 3;

    pointers = new QList<TouchPoint*>();

    TouchPoint *currentPointer = new TouchPoint();

    unsigned char firstByte = data.at(pointerDataOffset);
    unsigned char secondByte = data.at(pointerDataOffset + 1);
    int x = firstByte << 8 | secondByte;

    firstByte = data.at(pointerDataOffset + 2);
    secondByte = data.at(pointerDataOffset + 3);
    int y = firstByte << 8 | secondByte;

    QList<QPoint*> *locations = Screenshot::Instance()->getScreenBounds(Screenshot::Instance()->screenIndex);
    QPoint* startLocation = locations->at(0);
    QPoint* endLocation = locations->at(1);
    Size screenSize = {endLocation->x() - startLocation->x(), endLocation->y() - startLocation->y()};

    currentPointer->x = (x * screenSize.Width) / absoluteMaximum;
    currentPointer->y = (y * screenSize.Height) / absoluteMaximum;

    cursorPositionNew->setX(startLocation->x() + currentPointer->x);
    cursorPositionNew->setY(startLocation->y() + currentPointer->y);

    pointers->append(currentPointer);

    P1_Start->setX(currentPointer->x);
    P1_Start->setY(currentPointer->y);

    MouseMac::Instance()->moveMouseTo(cursorPositionNew->x(), cursorPositionNew->y());

    if (isPresenter)
    {
        updatePointerPosition();
    }
}

int MouseV3::trim(int value, int min, int max)
{
    return (value < min) ? min : ((value > max) ? max : value);
}

void MouseV3::updatePointerPosition()
{
    if (pointers->length() > 0)
    {
//        Server::Instance()->pointer.showPointer
        Logger::Instance()->add("Showing pointer");
        // todo
    }
}

void MouseV3::updateCursorPosition()
{
    if (!mousePadDown)
    {
        pointersDown();
        return;
    }
    P1_New = new QPoint(pointers->at(0)->x, pointers->at(0)->y);

    if ((P1_Start->x() == P_ORIGIN->x() && P1_Start->y() == P_ORIGIN->y()) || (P1_Last->x() == P_ORIGIN->x() && P1_Last->y() == P_ORIGIN->y()))
    {
        P1_Start = new QPoint(P1_New->x(), P1_New->y());
        P1_Last = new QPoint(P1_New->x(), P1_New->y());
        return;
    }

    cursorPositionCurrent = MouseMac::Instance()->getCursorPosition();

    if (Settings::Instance()->mouseAcceleration == 1.0)
    {
        P1_Rel->setX((P1_New->x() - P1_Last->x()) * Settings::Instance()->mouseSensitivity);
        P1_Rel->setY((P1_New->y() - P1_Last->y()) * Settings::Instance()->mouseSensitivity);
    }
    else
    {
        P1_Rel->setX(P1_New->x() - P1_Last->x());
        P1_Rel->setY(P1_New->y() - P1_Last->y());

        if (P1_Rel->x() != 0 || P1_Rel->y() != 0)
        {
            if (P1_Rel->x() > 0)
            {
                P1_Rel->setX(pow(P1_Rel->x(), Settings::Instance()->mouseAcceleration) * Settings::Instance()->mouseSensitivity);
            }
            else
            {
                P1_Rel->setX(P1_Rel->x() * -1);
                P1_Rel->setX(pow(P1_Rel->x(), Settings::Instance()->mouseAcceleration));
                P1_Rel->setX(P1_Rel->x() * Settings::Instance()->mouseSensitivity);
                P1_Rel->setX(P1_Rel->x() * -1);
            }

            if (P1_Rel->y() > 0)
            {
                P1_Rel->setY(pow(P1_Rel->y(), Settings::Instance()->mouseAcceleration) * Settings::Instance()->mouseSensitivity);
            }
            else
            {
                P1_Rel->setY(P1_Rel->y() * -1);
                P1_Rel->setY(pow(P1_Rel->y(), Settings::Instance()->mouseAcceleration));
                P1_Rel->setY(P1_Rel->y() * Settings::Instance()->mouseSensitivity);
                P1_Rel->setY(P1_Rel->y() * -1);
            }
        }
    }

    if (P1_Rel->x() < -700 || P1_Rel->x() > 700 || P1_Rel->y() < -700 || P1_Rel->y() > 700)
    {
        pointersUp();
        return;
    }

    P1_Last = new QPoint(P1_New->x(), P1_New->y());

    int newX = cursorPositionCurrent->x() + P1_Rel->x();
    int newY = cursorPositionCurrent->y() + P1_Rel->y();
    if (MouseMac::Instance()->moveMouseTo(newX, newY))
    {
        cursorPositionNew->setX(newX);
        cursorPositionNew->setY(newY);
    }
}

void MouseV3::parseMultitouch()
{
    if (!mousePadDown)
    {
        pointersDown();
        return;
    }

    P1_New = new QPoint(pointers->at(0)->x, pointers->at(0)->y);
    P2_New = new QPoint(pointers->at(1)->x, pointers->at(1)->y);

    if ((P1_Start->x() == P_ORIGIN->x() && P1_Start->y() == P_ORIGIN->y())
            || (P1_Last->x() == P_ORIGIN->x() && P1_Last->y() == P_ORIGIN->y())
            || (P2_Start->x() == P_ORIGIN->x() && P2_Start->y() == P_ORIGIN->y())
            || (P2_Last->x() == P_ORIGIN->x() && P2_Last->y() == P_ORIGIN->y()))
    {

        P1_Start = new QPoint(P1_New->x(), P1_New->y());
        P1_Last = new QPoint(P1_New->x(), P1_New->y());
        P2_Start = new QPoint(P2_New->x(), P2_New->y());
        P2_Last = new QPoint(P2_New->x(), P2_New->y());
        return;
    }

    P1_Rel->setX((P1_New->x() - P1_Last->x()) * Settings::Instance()->mouseSensitivity);
    P1_Rel->setY((P1_New->y() - P1_Last->y()) * Settings::Instance()->mouseSensitivity);

    if (P1_Rel->x() < -700 || P1_Rel->x() > 700 || P1_Rel->y() < -700 || P1_Rel->y() > 700)
    {
        pointersUp();
        return;
    }

    P1_Last = new QPoint(P1_New->x(), P1_New->y());
    P2_Last = new QPoint(P1_New->x(), P1_New->y());

    processMultitouch();
}

void MouseV3::processMultitouch()
{
    P3_New->setX(round((P1_New->x() + P2_New->x()) / 2.0));
    P3_New->setY(round((P1_New->y() + P2_New->y()) / 2.0));
    P3_Vector_New = Converter::Instance()->getPointDistance(*P1_New, *P2_New);

    if (P3_Start->x() == P_ORIGIN->x() && P3_Start->y() == P_ORIGIN->y())
    {
        P3_Start = new QPoint(P3_New->x(), P3_New->y());
    }

    if (P3_Vector_Start == 0)
    {
        P3_Vector_Start = P3_Vector_New;
    }

    if (currentGesture == GESTURE_NONE)
    {
        if (valueMatchesTolerance(P3_Vector_New, P3_Vector_Start, SCROLL_OFFSET_TOLERANCE))
        {
            if (valueMatchesTolerance(P3_New->x(), P3_Start->x(), SCROLL_OFFSET_TOLERANCE))
            {
                if (!valueMatchesTolerance(P3_New->y(), P3_Start->y(), SCROLL_OFFSET_TOLERANCE / 2))
                {
                    currentGesture = GESTURE_SCROLL;
                }
            }
        }
        else
        {
            currentGesture = GESTURE_ZOOM;
            P3_Vector_Event = P3_Vector_New + 20;
        }
    }

    if (currentGesture == GESTURE_ZOOM)
    {
        if (!valueMatchesTolerance(P3_Vector_New, P3_Vector_Event, 25))
        {
            scrollAmount = pow((P3_Vector_New = P3_Vector_Last), 2);
            if (P3_Vector_New > P3_Vector_Last)
            {
                MouseMac::Instance()->zoom(1, 1);
            }
            else
            {
                MouseMac::Instance()->zoom(-1, 1);
            }
            P3_Vector_Event = P3_Vector_New;
        }
    }
    else if (currentGesture == GESTURE_SCROLL)
    {
        scrollAmount = abs(pow(P3_New->y() - P3_Last->y(), 2));
        if (P3_New->y() > P3_Last->y())
        {
            MouseMac::Instance()->mouseScrollVertical(1, scrollAmount);
        }
        else if (P3_New->y() < P3_Last->y())
        {
            MouseMac::Instance()->mouseScrollVertical(-1, scrollAmount);
        }
    }

    P3_Last = new QPoint(P3_New->x(), P3_New->y());
    P3_Vector_Last = P3_Vector_New;
}

void MouseV3::checkForClick()
{
    long timeDelta = QDateTime::currentDateTime().toMSecsSinceEpoch() - pointDown->timestamp;

    if (timeDelta < CLICK_DELAY_TOLERANCE)
    {
        if (pointers->length() >= 1)
        {
            Logger::Instance()->add("CLick delay " + timeDelta);
            int offsetX = abs(P1_Start->x() - pointers->at(0)->x);
            int offsetY = abs(P1_Start->y() - pointers->at(0)->y);

            if (offsetX < CLICK_OFFSET_TOLERANCE && offsetY < CLICK_OFFSET_TOLERANCE)
            {
                leftSecondClick();
            }
        }
    }
}

bool MouseV3::valueMatchesTolerance(float val1, float val2, int tolerance)
{
    return (val1 < (val2 + tolerance) && val1 > (val2 - tolerance)) ? true : false;
}
